### PR TITLE
Adding `--iree-opt-import-parameters=` and cleaning up export.

### PR DESCRIPTION
**Pull request by benvanik**
The new option allows for importing all or some parameters from a parameter file into the compiler. One or more parameter files to one or more scopes can be specified, just as with the `--parameters=[scope=]file.*` flag in the runtime tool. If a scope is provided only parameters with that scope will be imported. If a parameter (optionally within a scope) is named explicitly it will be imported. If a maximum size is specified all parameters <= that size will be imported.

The export pass has been reworked to be the inverse of the import such that it's possible to round-trip through export and import. The import pass doesn't currently support all of the types we can export as `SerializableAttrInterface` doesn't have a deserialize method (yet) and we need special handling of packed types (i1, i4, etc). For now warnings are emitted and the globals are left as parameter references to still allow partial imports.

Flag renames:
* `--iree-opt-parameter-archive-export-file=` + `--iree-opt-parameter-archive-export-scope=` -> `--iree-opt-export-parameters=[scope=]file`
* `--iree-opt-minimum-parameter-export-size=` -> `--iree-opt-export-parameter-minimum-size=`
* `--iree-opt-splat-parameter-archive-export-file=` -> `--iree-opt-splat-parameters=`

### CODE REVIEW

File: compiler/src/iree/compiler/Modules/IO/Parameters/Transforms/test/import_parameters.mlir [hunk 1]
```diff
@@ -0,0 +1 @@
+// DO NOT SUBMIT
```
Comment (ScottTodd) [inline compiler/src/iree/compiler/Modules/IO/Parameters/Transforms/test/import_parameters.mlir]: Tagging this in a comment to resolve (add tests)

File: compiler/src/iree/compiler/Modules/IO/Parameters/Transforms/ImportParameters.cpp [hunk 2]
```diff
@@ -0,0 +1,419 @@
+// Copyright 2023 The IREE Authors
```
Comment (ScottTodd) [inline compiler/src/iree/compiler/Modules/IO/Parameters/Transforms/ImportParameters.cpp]: nit: if file is new, could use 2024

```suggestion
// Copyright 2024 The IREE Authors
```

File: compiler/src/iree/compiler/Modules/IO/Parameters/Transforms/ImportParameters.cpp [hunk 3]
```diff
@@ -0,0 +1,419 @@
...
+    return handleRuntimeError(
+        moduleOp, iree_io_parse_irpa_index(fileHandle->get(), parameterIndex),
+        "parsing irpa file");
+  } else if (format == ".safetensors") {
+    return handleRuntimeError(
+        moduleOp,
+        iree_io_parse_safetensors_index(fileHandle->get(), parameterIndex),
+        "parsing safetensors file");
+  } else {
+    llvm::errs() << "unsupported archive file format: " << path << "\n";
```
Comment (ScottTodd) [inline compiler/src/iree/compiler/Modules/IO/Parameters/Transforms/ImportParameters.cpp]: nit: wrap strings in quotes when logging them, so empty strings stand out
```suggestion
    llvm::errs() << "unsupported archive file format: '" << path << "'\n";
```

File: compiler/src/iree/compiler/Modules/IO/Parameters/Transforms/ImportParameters.cpp [hunk 4]
```diff
@@ -0,0 +1,419 @@
...
+    auto *parameterIndex = parameterIndices.lookupOrCreate(moduleOp, scope);
+    if (failed(loadParameterIndex(moduleOp, path, parameterIndex)))
+      return failure();
+  }
+  return parameterIndices;
+}
+
+// Today only shaped types of elements where we know we can directly access the
+// data as stored in teh file.
+static bool isTypeSupported(Type type) {
```
Comment (ScottTodd) [inline compiler/src/iree/compiler/Modules/IO/Parameters/Transforms/ImportParameters.cpp]: nit: typo
```suggestion
// Today only shaped types of elements where we know we can directly access the
// data as stored in the file.
static bool isTypeSupported(Type type) {
```

**Review by ScottTodd — APPROVED** [review]

File: compiler/src/iree/compiler/Modules/IO/Parameters/Transforms/ImportParameters.cpp [hunk 5]
```diff
@@ -0,0 +1,419 @@
...
+      value = llvm::support::endian::read32le(entry->storage.splat.pattern);
+      break;
+    case 64:
+      value = llvm::support::endian::read64le(entry->storage.splat.pattern);
+      break;
+    default:
+      assert(false && "integer width not supported");
+      return failure();
+    }
+    return IntegerAttr::get(globalType, APInt(integerType.getWidth(), value));
```
Comment (ScottTodd) [inline compiler/src/iree/compiler/Modules/IO/Parameters/Transforms/ImportParameters.cpp]: build errors around here
https://github.com/openxla/iree/actions/runs/8345780110/job/22841503854?pr=16828#step:4:7622
```
 error: no viable conversion from returned value of type 'mlir::IntegerAttr' to function return type 'FailureOr<mlir::TypedAttr>'
    return IntegerAttr::get(globalType, APInt(integerType.getWidth(), value));
           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by benvanik** [comment]
In order to get a satisfying test for import I had to clean up export; PTAL!

File: runtime/src/iree/io/formats/parser_registry.h [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+#include "iree/base/api.h"
+#include "iree/io/file_handle.h"
+#include "iree/io/parameter_index.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif  // __cplusplus
+
+// DO NOT SUBMIT path may be full path or just extension
```
Comment (ScottTodd) [inline runtime/src/iree/io/formats/parser_registry.h]: 👀

**Review by ScottTodd — COMMENTED** [review]
reviewing

File: compiler/src/iree/compiler/Modules/IO/Parameters/Transforms/Passes.td [hunk 7]
```diff
@@ -11,32 +11,49 @@ include "mlir/Pass/PassBase.td"
...
+    Option<"filePath", "file", "std::string",
            /*default=*/"",
-           "Path to write the parameter archive to.">
+           "Path to write the parameter archive to.">,
+  ];
+}
+
+def ImportParametersPass :
+    Pass<"iree-io-import-parameters", "mlir::ModuleOp"> {
+  let summary = "Imports parameters from an archive file.";
```
Comment (ScottTodd) [inline compiler/src/iree/compiler/Modules/IO/Parameters/Transforms/Passes.td:43]: We should also update https://iree.dev/guides/parameters/#working-with-parameter-files

Guess I can add that to my docs backlog along with updating the profiling / Tracy docs... and soon maybe also the PyTorch docs...

File: compiler/src/iree/compiler/Modules/IO/Parameters/Transforms/test/import_parameters.mlir [hunk 8]
```diff
@@ -0,0 +1,36 @@
+// RUN: iree-opt --pass-pipeline="builtin.module(iree-io-export-parameters{path="opt=%t.irpa" minimum-size=0},iree-io-import-parameters{paths="opt=%t.irpa"})" %s | FileCheck %s
```
Comment (ScottTodd) [inline compiler/src/iree/compiler/Modules/IO/Parameters/Transforms/test/import_parameters.mlir:1]: Checking my understanding: this exports _then_ imports, round tripping?

I suppose we could also test with `iree-create-parameters`, if we didn't want to rely on the sequencing there. Feels sneaky to me to have the files needed by the compiler not all already available on disk when the command begins.

**Review by ScottTodd — COMMENTED** [review]
Code changes LGTM. A few CI failures still + one `DO NOT SUBMIT` in a comment

File: runtime/src/iree/io/formats/parser_registry.h (hunk 6 above)
Comment (benvanik) [reply runtime/src/iree/io/formats/parser_registry.h]: whoops

File: compiler/src/iree/compiler/Modules/IO/Parameters/Transforms/test/import_parameters.mlir (hunk 8 above)
Comment (benvanik) [reply compiler/src/iree/compiler/Modules/IO/Parameters/Transforms/test/import_parameters.mlir:1]: round-tripping is best - it's more sneaky to have runtime deps in compiler lit tests :)

Comment (ScottTodd) [reply compiler/src/iree/compiler/Modules/IO/Parameters/Transforms/test/import_parameters.mlir:1]: Ah, good point. SGTM

**Comment by benvanik** [comment]
This exposed some real nastiness in our global opt pipeline around constant hoisting and inlining that the export pass was previously working around by running as part of the fixed point iteration in just the right place. I'm out of brain today but will see if I can think of something robust next week - if not, I'll make a flag that forces hoisting at a deterministic point for the export pass.

**Comment by stellaraccident** [comment]
I can help next week. Out of brain for today. But thank you for working on it.

File: compiler/src/iree/compiler/Modules/IO/Parameters/Transforms/Passes.td (hunk 7 above)
Comment (ScottTodd) [reply compiler/src/iree/compiler/Modules/IO/Parameters/Transforms/Passes.td:43]: I could fold doc updates into https://github.com/openxla/iree/pull/17006 if this PR lands first

**Comment by ScottTodd** [comment]
Nice, so you fixed what was blocking this? Is the new import flag now ready to be documented on https://iree.dev/guides/parameters/ (together with the export flag)?

**Pull request merged**